### ISSUE TITLE
Printing and showing ITensor in non-transposed format

### ISSUE THREAD

**Issue by jtschneider** [ITensor/ITensors.jl#154]
In view of closed issue https://github.com/ITensor/ITensors.jl/issues/152, I propose a change in the printing and passing mechanism of an ITensor.
The issue is briefly summarised with the following code example:

```julia
julia> using ITensors

julia> sites = ITensors.siteinds("S=1/2", 1)
1-element Array{Index,1}:
 (2|id=90|S=1/2,Site,n=1)

julia> Sy = op(sites[1], "Sy");

julia> @show Sy
Sy = ITensor ord=2 (2|id=90|S=1/2,Site,n=1) (2|id=90|S=1/2,Site,n=1)'
Dense{Complex{Float64},Array{Complex{Float64},1}}
2×2 Tensor{Complex{Float64},2,Dense{Complex{Float64},Array{Complex{Float64},1}},IndexSet{2}}:
  0.0+0.0im  0.0+0.5im
 -0.0-0.5im  0.0+0.0im
ITensor ord=2 (2|id=90|S=1/2,Site,n=1) (2|id=90|S=1/2,Site,n=1)'
Dense{Complex{Float64},Array{Complex{Float64},1}}

julia> matrix(Sy)
2×2 Array{Complex{Float64},2}:
  0.0+0.0im  0.0+0.5im
 -0.0-0.5im  0.0+0.0im
```

`@show {ITensor}` and `matrix({ITensor})` prints and returns, respectively, the transposed version of a conventional matrix, as it begins with the unprimed (ket) index and then the primed (bra) index, rather than the other way around. Indeed, thinking of `Sy` as a matrix mapping (a vector to its left and a dualvector to is right) to the complex numbers, would expect the indices to be permuted.
This has proven to be confusing in at least one case 🙋‍♂️.

What are your thoughts?

Cheers,
Jan

**Comment by jtschneider** [comment]
Addendum:
For the sake of completeness and to immediately see why I was confused, here is the definition of `Sy` for `S=1/2`:

<img width="210" alt="image" src="https://user-images.githubusercontent.com/9451057/70319160-eb5dc580-1821-11ea-9c41-4a3985bdce54.png">

**Comment by orialb** [comment]
> @show {ITensor} and matrix({ITensor}) prints and returns, respectively, the transposed version of a conventional matrix, as it begins with the unprimed (ket) index and then the primed (bra) index, rather than the other way around. 

Maybe it would make sense to not allow calling `matrix(T::ITensor)` without specifying the indices, such that only `matrix(T::ITensor, Li::Index, Ri::Index)` (maybe without Ri) would be allowed? 
This would seem to fit more the exisitng API for linear algebra operations on ITensors and to the general philosophy that ITensors don't have a notion of index ordering.

**Comment by emstoudenmire** [comment]
Jan, Ori, thanks this is a really useful discussion.

Ori: I agree perhaps we should remove the `matrix(T::ITensor)` function altogether. I think it's currently intended as a lower-level / advanced functionality, though, like for use inside of other "interface/user level" ITensor functions where one wants to convert various storage types into a dense matrix for various internal purposes. So we may alternatively end up keeping it and just emphasizing through our documentation that users prefer the `matrix(T::ITensor, i1::Index, i2::Index)` function instead, which already exists too.

Jan: while I think the idea of making the printing automatically show the "right" order could be a good one, I do have some concerns about it. One is that, in general, there is no known correct ordering for general tensors. Consider a tensor where the indices have a range of prime levels, and some have the same prime level and id, yet different tags. For all of those cases we wouldn't have a preferred ordering.  

On the other hand, we may want to revisit this for the case of at least two-index tensors, because of some upcoming features I want to add regarding fermions. Within this upcoming approach, if the code printed the density "n" operator with the indices reversed, it would show the matrix element which is normally written as +1 as -1 instead! So I think users could be pretty surprised that their operators have the "wrong" sign (even though this wouldn't be a bug). 

But if we only do this sort of treatment for two-index tensors, then we have a situation where sometimes we do one thing, and sometimes another. Plus it inhibits debugging of code, where printing is useful to understand things like the way ITensors are laid out in memory.

So I think for now I will just start by changing operators to default to the usual row, column ordering where they are defined (in files like src/physics/site_types/spinhalf.jl) so that the current default printing scheme shows the usual expected ordering.

Best regards,
Miles

**Comment by orialb** [comment]
> I think it's currently intended as a lower-level / advanced functionality, though, like for use inside of other "interface/user level" ITensor functions where one wants to convert various storage types into a dense matrix for various internal purposes. So we may alternatively end up keeping it and just emphasizing through our documentation that users prefer the matrix(T::ITensor, i1::Index, i2::Index) function instead, which already exists too.

I agree that this can be solved at the documentation level. Maybe one just has to make sure that the docstring for `matrix(T::ITensor{N}, i1::Index, i2::Index)` comes up before `matrxi(T::ITensor)` when calling `? matrix` in the REPL (once there are docstrings :) ).

**Comment by emstoudenmire** [comment]
Thanks Ori - good suggestion. I just added a docstring which follows your suggestion.

I'm going to close this issue since I think the manual ordering of the operator indices (done in the commit referenced above) is the best solution I know of right now to this issue, I think.

Thanks again Jan for pointing this out.

**Comment by mtfishman** [comment]
I agree with Miles that the `matrix(::ITensor)` function is meant as a more advanced, developer-level function. The idea behind the `matrix(::ITensor)` function (or more generally the `array(::ITensor)`) is that if the ITensor has a data structure that can be mapped to a dense array directly, it creates a "view" of that data, i.e.:

```julia
i,j = Index(2),Index(2)
T = randomITensor(i,j)
A = array(T)
A[1,1] = 1.0   # Also modifies T
T[1,1] == 1.0
```

If it doesn't have a "dense" data format, such as a Diag ITensor, it creates the closest `Array` to the ITensor (of course not as a view).

This in general is a microptimization, but is useful in functions when you want to avoid extra copies (in cases when you know the ordering doesn't need to be changed). Right now it is mainly used in testing Dense ITensor functionality, where we want to switch back and forth between ITensors and Julia Arrays to test that ITensor functions are working correctly.

I actually meant to name functions that take the Index order `Array(::ITensor, ::Index...)` instead of the lower-case `array(...)` to make it clear that they will always make a copy (since in general it would require permuting the ITensor data). This is a general Julia convention that constructors should be "owning" (which is not always the case, like creating an ITensor from a vector of data, but I think is a good convention in this case). It would also be nice to have versions that explicitly takes type parameters like `Array{T}(::ITensor, ::Index...)` as well, to help with type inference.

So tldr: `array(::ITensor)` is meant as an internal function, and the version explicitly taking indices should be renamed `Array(::ITensor, ::Index...)` (and `Matrix(::ITensor,::Index,::Index)`) for clarity. In general, users should not use `array`, but should use `Array` and related functions.

**Comment by mtfishman** [comment]
Also, I think it would be good to have a version of `print` (or whatever that relevant function is to overload) where you can specify the ordering of the indices that the ITensor will be printed with, since I think the main application of specifying the ordering of the ITensor indices (in my experience) is for printing. An example would be a custom printing macro for ITensors like:

```julia
A = randomITensor(i,j)
@showit A[j,i]   # prints A with indices ordered as j,i
```

This will help in cases beyond the one mentioned in this issue that can be handled by changing the internal convention for the constructor of ITensor operators.

Note that you can do something similar right now with the `permute` function to specify index ordering, which can help with printing in a specified order:

```julia
julia> i,j = Index(2,"i"),Index(2,"j")
((2|id=751|i), (2|id=377|j))

julia> T = randomITensor(i,j)
ITensor ord=2 (2|id=751|i) (2|id=377|j) 
Dense{Float64,Array{Float64,1}}

julia> @show T;
T = ITensor ord=2 (2|id=751|i) (2|id=377|j) 
Dense{Float64,Array{Float64,1}}
Tensor{Float64,2,Dense{Float64,Array{Float64,1}},IndexSet{2}}
 2×2
 0.6937405073189952  -2.0122587718427094
 1.5835387883013172  -1.6791743195482187

julia> @show permute(T,j,i);
permute(T, j, i) = ITensor ord=2 (2|id=377|j) (2|id=751|i) 
Dense{Float64,Array{Float64,1}}
Tensor{Float64,2,Dense{Float64,Array{Float64,1}},IndexSet{2}}
 2×2
  0.6937405073189952   1.5835387883013172
 -2.0122587718427094  -1.6791743195482187
```

**Comment by orialb** [comment]
Thanks for the clarifications Matt. 
Your explanation regarding the `matrix` function acting differently depending on the storage type made me think of another off-topic question regarding the design- 
What is the advantage of not having the storage type as a parameter of the ITensor type (`ITensor{N,S<:TensorStorage}`) ? Wouldn't that also allow better type inference? Or is it due to the fact that all the actual work is done at the `Tensor` level or lower, that there is no performance penalty for not parametrizing `ITensor`?  
I'm asking just out of curiosity and to better understand the design principles, as I'm sure you thought about those things.

**Comment by orialb** [comment]
I guess the following section from the Julia docs answers my question- https://docs.julialang.org/en/latest/manual/performance-tips/#kernel-functions-1 , where the key is "Julia's compiler specializes code for argument types at function boundaries" (which I have read before, somehow didn't think of it in this context).

**Comment by mtfishman** [comment]
This is a commonly asked question, we will make a write-up about it eventually.

The advantage of not parametrizing by the storage type is to allow for more generic in-place code. For example, this works in ITensors.jl:
```julia
julia> i,j = Index(2,"i"),Index(2,"j")
((2|id=878|i), (2|id=160|j))

julia> T = randomITensor(i,j)
ITensor ord=2 (2|id=878|i) (2|id=160|j) 
Dense{Float64,Array{Float64,1}}

julia> T1 = randomITensor(i,j)
ITensor ord=2 (2|id=878|i) (2|id=160|j) 
Dense{Float64,Array{Float64,1}}

julia> T2 = randomITensor(ComplexF64,i,j)
ITensor ord=2 (2|id=878|i) (2|id=160|j) 
Dense{Complex{Float64},Array{Complex{Float64},1}}

julia> add!(T1,T2)
ITensor ord=2 (2|id=878|i) (2|id=160|j) 
Dense{Complex{Float64},Array{Complex{Float64},1}}

julia> T1
ITensor ord=2 (2|id=878|i) (2|id=160|j) 
Dense{Complex{Float64},Array{Complex{Float64},1}}
```
which could not work if the ITensors were parametrized by the storage type (and also works because ITensor is mutable). In general, this allows for nicer in-place operations where we can internally handle type conversions without the users having to worry about it (like imagine similar situations of adding Dense ITensors to Diag ITensors, it is nice to not have to think about the conversion).

Indeed, the idea is that using function barriers, the only overhead of calling ITensor functions over the internal Tensor functions should be the higher level index logic and a runtime dispatch to choose the correct Tensor function to call (this is part of why the Tensor layer exists, though it is also useful for code organization).

**Comment by orialb** [comment]
Thanks for the detailed reply!
